File: ga/gpt3_5_result_2024_6/result/1968G2_cpp_gpt/40_1968G2.cpp
use std::io::{self, BufRead};

const N: usize = 2_00005;

fn find(k: usize, p: &mut [usize; N], n: usize) -> usize {
    if k > n || p[k] == k {
        k
    } else {
        p[k] = find(p[k], p, n);
        p[k]
    }
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines().map(|line| line.unwrap());
    
    let t: usize = lines.next().unwrap().trim().parse().unwrap();
    
    for _ in 0..t {
        let line = lines.next().unwrap();
        let mut parts = line.split_whitespace();
        let n: usize = parts.next().unwrap().parse().unwrap();
        let x: usize = parts.next().unwrap().parse().unwrap();
        let y: usize = parts.next().unwrap().parse().unwrap();
        let a: Vec<char> = parts.next().unwrap().chars().collect();
        
        let mut p: [usize; N] = [0; N];
        let mut t: [usize; N] = [0; N];
        let mut ans: [usize; N] = [0; N];
        let mut z: [usize; N] = [0; N];
        
        z[1] = n;
        let mut i = 2;
        let mut j = 0;
        
        while i <= n {
            if j + z[j] > i {
                z[i] = std::cmp::min(z[i - j + 1], j + z[j] - i);
            }
            
            let mut k = i + z[i];
            while a[k] == a[1 + z[i]] {
                z[i] += 1;
                k += 1;
            }
            
            if i + z[i] > j + z[j] {
                j = i;
            }
            
            i += 1;
        }
        
        let mut t_sorted: Vec<usize> = (1..=n).collect();
        t_sorted.sort_by_key(|&i| z[i]);
        
        let mut k = 1;
        
        for i in 1..=n {
            let mut c = 0;
            while k <= n && z[t_sorted[k]] < i {
                p[t_sorted[k]] = t_sorted[k] + 1;
                k += 1;
            }
            
            let mut j = 1;
            while j <= n {
                j = find(j + i, &mut p, n);
                c += 1;
            }
            
            ans[c] = i;
        }
        
        for i in (1..=n).rev() {
            ans[i] = std::cmp::max(ans[i], ans[i + 1]);
        }
        
        for i in x..=y {
            print!("{} ", ans[i]);
        }
        println!();
    }
}

File: ga/glm4_result_2024_6/result/1968G2_cpp_glm/40_1968G2.rs
use std::io;
use std::cmp::min;
use std::collections::HashMap;

const N: usize = 2e5 as usize + 3;
fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let t: usize = input.trim().parse().unwrap();
    
    for _ in 0..t {
        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap();
        let mut tokens = input.trim().split_whitespace();
        let n: usize = tokens.next().unwrap().parse().unwrap();
        let x: usize = tokens.next().unwrap().parse().unwrap();
        let y: usize = tokens.next().unwrap().parse().unwrap();
        let a: Vec<char> = tokens.next().unwrap().chars().collect();
        
        let mut p = vec![0; N];
        let mut t = vec![0; N];
        let mut ans = vec![0; N];
        let mut z = vec![0; N];
        
        for i in 1..=n {
            p[i] = i;
            t[i] = i;
            ans[i] = 0;
            z[i] = 0;
        }
        z[1] = n;
        
        for i in 2..=n {
            let mut j = 0;
            if j + z[j] > i {
                z[i] = min(z[i - j + 1], j + z[j] - i);
            }
            while a[i + z[i]] == a[1 + z[i]] {
                z[i] += 1;
            }
            if i + z[i] > j + z[j] {
                j = i;
            }
        }
        
        t[1..=n].sort_by(|&i, &j| z[i].cmp(&z[j]));
        
        for i in 1..=n {
            let mut k = 1;
            while k <= n && z[t[k]] < i {
                p[t[k]] = t[k] + 1;
                k += 1;
            }
            let mut c = 0;
            for j in 1..=n {
                let mut j = j;
                while j <= n {
                    c += 1;
                    j = p[j];
                }
            }
            ans[c] = i;
        }
        
        for i in (0..n).rev() {
            ans[i] = max(ans[i], ans[i + 1]);
        }
        
        for i in x..=y {
            print!("{} ", ans[i]);
        }
        println!("");
    }
}